### PR TITLE
Use interfaces for file and directory I/O operations. Connected to #69

### PR DESCRIPTION
**Pull request by anders9ustafsson**
Everyone, here is one *massive* pull request with practically no functionality changes at all :-)

I have refactored all file and directory based I/O to be interface-based. In the regular code, instead of calling methods e.g. from `System.IO.File`, `System.IO.DirectoryInfo`, `System.IO.Path`, all these operations are accessed via `Dicom.IO.IOManager`.

`IOManager` is an abstract class, and it publicly exposes the following static properties and methods:

    IPath Path { get; }
    IFileReference CreateFileReference(string fileName);
    IDirectoryReference CreateDirectoryReference(string directoryName);

Similar to `LogManager`, `IOManager` holds an `IOManager` derived implementation object for actually performing the requested I/O operations.  For the current .NET based version of *fo-dicom*, I have implemented `DesktopIOManager` that makes use of the `DesktopPath`, `DesktopFileReference` and `DesktopDirectoryReference` implementations, respectively. If (when!) *fo-dicom* is transferred to other platforms, we can simply implement platform specific `IOManager`, `IFileReference` etc. classes to provide corresponding I/O functionality on these platforms.

Please also note that `DesktopFileReference` is really the "old" `FileReference` renamed and extended. So there was already a seed for implementing this sort of platform-agnostic file I/O management; I have now gone about and implemented it in full.

Once we move *fo-dicom* over to portable class libraries, this refactoring will be of immense help. The abstract `IOManager` and the new interfaces remain in the fully portable core library, whereas the `Desktop`... implementations are moved to a .NET specific support library. Since all *fo-dicom* code with this refactoring rely on I/O interfaces, there will be no problem maintaining this code in the portable core library.

The public API of *fo-dicom*  has not changed at all from this refactoring. All *fileName* and *directoryName* methods remain intact.

For this I/O interface approach to continue to work, it will require some discipline from the *fo-dicom* contributors: 
* For all file- and directory-based I/O, use `IOManager`. 
* If a specific I/O operation is missing from the interfaces, add it to the relevant interface and implement it in the corresponding `Desktop`... class. 
* DO NOT use the following classes from the `System.IO` namespace (except in the `Desktop`... I/O interface implementations): `File`, `Directory`, `FileInfo`, `DirectoryInfo`, `FileStream`, `Path`.

In conclusion, this PR does not change anything in the *fo-dicom*  behavior or public API, but internally it considerably simplifies a transfer of *fo-dicom* to other target platforms via a Portable Class Library approach. Personally I consider this to be a crucial step going forward, eventually enabling us to target the new *Universal Windows Platform* as well as *Android* and *iOS* via Xamarin.

Please review this pull request thoroughly and share your thoughts and concerns.

### CODE REVIEW

**Comment by anders9ustafsson** [comment]
OK, apparently the unit tests fail to run on the CI build machine due to Stack Overflow exception. The code builds without problem, but I will check if I can get rid of this unit test error as soon as possible.

**Comment by anders9ustafsson** [comment]
OK, fixed!

To a large extent, I have added unit tests in those areas where I have made code modifications, and I have not seen any issues with the changes I have made (except for the initially failing unit tests, which was due to a recursion error in temporary file removal that I had introduced, fixed now).

I have also taken the opportunity to add API documentation to at least some of the methods, properties and types that I have refactored. Hence less compilation warnings in Release mode... :-)

I should add that as part of the refactoring to avoid platform specific file I/O operations, I have removed the `DicomFile.OpenMemoryMapped` method(s) that I introduced in #71 . The same effect should be accomplished by calling `DicomFile.Open(Stream)`, where the stream is a view stream in a `MemoryMappedFile`.

**Comment by IanYates** [comment]
Wow - great work here.  I'm pretty heads down for about a week (I've got a conference to prepare for and then attend) but will keep up on the chatter here.  I'll see if I can at least read through the code and run it.
Regarding #71, so the path forward for memory mapped files, if someone wishes to use them, is to use the Stream-based overloads, correct?  #71 has been merged so I guess this PR effectively undoes some or all of it.

**Comment by anders9ustafsson** [comment]
@IanYates Good luck with the conference! 

In the end the `DicomFile.OpenMemoryMapped` addition was only a tiny part of #71. The main concern of #71 was the reversal of `PixelData` classes to the pre-1.1.0 approach. 

The `Stream`-based `DicomFile.Open` overloads effectively mimic `DicomFile.OpenMemoryMapped` when the stream is obtained from a `MemoryMappedFile`, so there should be no need for `MemoryMappedFile` specific methods. I have added a unit test to illustrates that approach, [here](https://github.com/fo-dicom/fo-dicom/pull/78/files#diff-acdfa09b156bc9bf13d57d99eedb48c2R111).

**Comment by hiscodeness** [comment]
I browsed the changes and they look good to me. Or at least didn't come across anything that I would be concerned about. Providing overloads with Streams is great :+1:.

I didn't compile or run the tests, as this is currently not an important feature for me. Sorry about being so lazy. But I'm all for abstracting any platform specific stuff so the PR gets an OK from me.

**Comment by anders9ustafsson** [comment]
Thanks, Jussi, it's very much appreciated that you took the time to review the code! If no concerns are raised in the near future, I plan to merge this PR fairly soon.
